### PR TITLE
Make page_count self-truncating when we detect the last page of a query

### LINKED ISSUE #196: Dead link filtering reduces number of queryable pages
## Description

The dead link filtering algorithm significantly limits the pages that can be retrieved.

## Reproduction

0. Go to audio search endpoint: `http://localhost:8000/v1/audio/`. Note the number of pages.
1. Go to a page number that's quite high but less than 250, like 200: `http://localhost:8000/v1/audio/?page=200`
2. See the deep pagination error. This should not be happening.

## Expectation

Page count calculation should be correct and should allow pagination upto the stated limit.

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #196  by @dhruvkb 

## Description

This relatively simple change "truncates" the page count for a query to the current page number if we receive fewer results on the page than are able to fill out the page based on the page_size. This follows the suggestion from the issue by @AetherUnbound.

## Testing Instructions

Checkout the branch and run the project according to the README.

Test the result views and pay attention to the page count. Test queries that have a single page, multiple pages, 20 pages, etc. Tip: use increasingly specific terms to get smaller returns from the test data.

Review the unit test cases added and confirm they are sensible.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by sarayourfriend** [comment]
I am debugging the tests that are failing. It is easy to fix them, I just change the page_count from 0 to 1 to match the new expected result of the page counting operation... however, now I am second guessing whether `page_count` should be 1 or 0 when there is only a single page of results. Do any @WordPress/openverse-api folks know what the intention is? My intuition is that it should be 1 as there is a single page of results. I've made changes to reflect that, but again, I'm second guessing myself that I actually understand the original intention of the code, in particular because the "dangling page of results" code was obscure to me (why did we modulo page count over page size? this made no sense to me).

Thanks for any insight here.

**Comment by zackkrida** [comment]
This was discussed a bit in the community meeting in the [Make WP Slack](https://make.wordpress.org/chat/):

https://wordpress.slack.com/archives/C02012JB00N/p1670945242644929

@obulat shared that the frontend expects the page count to be `1`, and I mentioned that an api response with these values:

```json
{
  "page": 1,
  "page_count": 1
}
```

feels internally consistent. 

Additionally, the API currently requires users to pass a `page` value of 1 or higher:

https://api.openverse.engineering/v1/audio/?q=puppy&page=0

```json
{
    "detail": {
        "page": [
            "Ensure this value is greater than or equal to 1."
        ]
    }
}
```

**Comment by sarayourfriend** [comment]
That is all in line with what my expectations would be as well. Thanks for uncovering those details, Zack. I'll undraft this now :slightly_smiling_face:

**Comment by sarayourfriend** [comment]
I will be AFK starting tomorrow so Dhruv and Olga, if y'all review and approve this feel free to merge it at your convenience if you're working.

File: api/catalog/api/controllers/search_controller.py
Comment (obulat) [inline api/catalog/api/controllers/search_controller.py:519]: I was afraid that this change will mean that we have to update the frontend code to account for the `page_count` changing between request, but we actually do update it on every request, so it's fine :)

File: api/test/unit/controllers/test_search_controller.py
Comment (obulat) [inline api/test/unit/controllers/test_search_controller.py:22]: In practice, this function can never be called with 0 as the `page_size` value because we have a validator on `page_size` with a minimum value of 1.

File: api/test/unit/controllers/test_search_controller.py
Comment (obulat) [inline api/test/unit/controllers/test_search_controller.py:25]: I don't understand why the page_count returned here is 0. Isn't it supposed to be 1 even if we have 1 result?

File: api/test/unit/controllers/test_search_controller.py
Comment (sarayourfriend) [reply api/test/unit/controllers/test_search_controller.py:22]: Correct, but unit tests generally are not meant to embed meta-knowledge of the application like that. The list here tests the unit's possible (relevant and irrelevant to the application) inputs which makes it easier to verify the intended behaviour (at least in my experience).

File: api/test/unit/controllers/test_search_controller.py
Comment (sarayourfriend) [reply api/test/unit/controllers/test_search_controller.py:25]: It should be, yes. I suppose it should only ever return 0 if there are 0 results. I'll look into this, thanks for catching it.

Comment (sarayourfriend) [reply api/test/unit/controllers/test_search_controller.py:25]: This turned out to mostly be a test quirk, as the page sent to the method was `0` even though that's not a real page that ever really gets requested because we do not 0 index the pages. However, it's easy to correct for it by treating `page = 0` the same as `page = 1`.

**Comment by sarayourfriend** [comment]
@obulat Can you re-review this? @dhruvkb when you have a chance this needs a second review. Thanks@

File: api/test/unit/controllers/test_search_controller.py
Comment (obulat) [reply api/test/unit/controllers/test_search_controller.py]: Thank you for looking into this, @sarayourfriend! Do you think a comment about it would be useful in the tests?

File: api/catalog/api/controllers/search_controller.py
Comment (obulat) [inline api/catalog/api/controllers/search_controller.py:512]: Honestly, this even seems redundant since we have the validators that prevent the users from ever setting the `page` to 0, and if you send the request with `page=0`, you get a 400 response: https://api.openverse.engineering/v1/images/?q=nature&page=0.

File: api/test/unit/controllers/test_search_controller.py
Comment (obulat) [inline api/test/unit/controllers/test_search_controller.py:40]: I guess we can remove the comment above now?

**Review by obulat — APPROVED** [review]
Fantastic test cases! 
I think one comment is redundant now that these changes have been introduced, note in the inline comments.

File: api/catalog/api/controllers/search_controller.py
Comment (sarayourfriend) [reply api/catalog/api/controllers/search_controller.py:512]: That's true. I had mixed feelings about it. Do you think I should instead update the tests to only pass inputs we know would be valid according to the API? That doesn't seem unreasonable, really...

Comment (obulat) [reply api/catalog/api/controllers/search_controller.py:512]: If we leave it as is, then it'll be very easy to allow for `page=0` in the API request in the future if we ever decide to do that. However, it will also mean that an unnecessary clause will run for every request. Even though it's very (extremely) short. I'm also not sure, but since we're not likely to change the allowed pages,, it seems better to save a tiny amount of resources on each request :)

Comment (sarayourfriend) [reply api/catalog/api/controllers/search_controller.py]: I changed it to remove the impossible edge case, thanks for the feedback :pray:

File: api/test/unit/controllers/test_search_controller.py
Comment (sarayourfriend) [reply api/test/unit/controllers/test_search_controller.py]: Done in https://github.com/WordPress/openverse-api/pull/1032/commits/bfa0802d00cdc4282e2063b8f91fac8e66f1ff9a

**Review by dhruvkb — APPROVED** [review]
The change makes sense and the tests are very comprehensive. If the frontend automatically works with it, that's icing on the cake!

**Pull request merged**
